### PR TITLE
WorldGen docs

### PR DESCRIPTION
**Pull request by Inscrutable**
This is but a start. Eventually this will fix issue #159.
Many thanks to @Deamon5550.
Things yet to be done:
- [ ] Check code snippets actually work
- [x] Split into separate pages (if necessary)
- [x] Fill in [TBA] and {Citation Needed}
- [x] Proofread and minimise text

### CODE REVIEW

**Comment by Spongy** [comment]
A preview for this pull request is available at https://cdn.rawgit.com/Spongy/SpongeDocs-PRs/e2e7d395/index.html.

Here are some links to the pages that were modified:

- added https://cdn.rawgit.com/Spongy/SpongeDocs-PRs/e2e7d395/plugin/wgen-index.html

_Since the preview frequently changes, please link to [this comment](https://github.com/SpongePowered/SpongeDocs/pull/368#issuecomment-154771700), not to the direct url to the preview._

File: source/plugin/wgen-index.rst [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+.. Note::
+	Block placement in the generation phase takes place entirely within a buffer,
+	as the Chunk object is not created until the last step in the phase.
+
+In the beginning, there is nothing, and then the base generator populator creates the basic shape of the terrain.
+At this stage the world is entirely made of stone. Default generation is simply a heightmap generated from 
+perlin noise, then filled in - leaving air if higher than sea level, and water if below. This step can be overridden
+by calling ``setBaseGeneratorPopulator`` to provide your own custom populator to the world generator.
+
+Next, the biome generator populates a 2d biome buffer with the biomes that will correspond to each 1x1 column
```
Comment (Deamon5550) [inline source/plugin/wgen-index.rst]: Just noticed these first two pieces are in the wrong order, The biomes are generated first as then the base terrain is then created based off of the information from the biomes. So the order should follow:

1. The BiomeGenerator populates a 2d buffer
2. The base GenerationPopulator creates the base terrain shape
3. the top layers of terrain are replaced according to the GroundCoverLayers
4. Etc.

File: source/plugin/wgen-index.rst [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+The *Population Phase* can only run on a chunk once the three chunks adjacent to it in the positive x and z axes
+are loaded. This allows objects during this phase to expand outside the 16x16 chunk area being populated, without
+requiring the partial generation used for ``GeneratorPopulator``\ s. To support this, the actual area populated
+during this phase is a 16x16 area offset by 8 in each of the x and z axes.
+
+[Image TBA ?]
+
+At this time the populators defined globally and those defined for the biome specified for the centre block of
+the populating region (which is the maximal block of the chunk passed to the populator due to the offset
+specified above). ???
```
Comment (Deamon5550) [inline source/plugin/wgen-index.rst]: What I was trying to saw was that only the populators of the biome at the position `(x*16+16, 0, z*16+16)` are applied to this area. Not a union of all the biomes as is the case for  GenerationPopulators

**Comment by Inscrutable** [comment]
I'm just drafting the rest up, trying to decide how much to keep and how many pages we split it into. Some feedback from docs staff and others would be useful. I will try to put up what I have done soon as a single page so we can see a preview, and go from there. This has many bits of [WIP] still in it.

File: source/plugin/wgen/customwgen.rst [hunk 3]
```diff
@@ -0,0 +1,251 @@
...
+
+[Insert an example or two.]
+
+Creating Custom Biomes
+~~~~~~~~~~~~~~~~~~~~~~
+
+*At present it is not possible to create entirely new biomes from purely within the SpongeAPI.*
+
+
+[TODO]
```
Comment (Deamon5550) [inline source/plugin/wgen/customwgen.rst]: Oh heh this was my todo for the world gen branch that somehow made it into that gist.

File: source/plugin/wgen/customwgen.rst [hunk 4]
```diff
@@ -0,0 +1,240 @@
...
+
+Creating a WorldGeneratorModifer
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Let's start with the format of a ``WorldGeneratorModifier``. First, you need a class which implements
+the WorldGeneratorModifier interface:
+
+.. code-block:: java
+
+	public class MyModifier implements WorldGeneratorModifier {
```
Comment (phroa) [inline source/plugin/wgen/customwgen.rst]: indentation here isn't consistent

File: source/plugin/wgen/customwgen.rst [hunk 5]
```diff
@@ -0,0 +1,240 @@
...
+
+    @Override
+    public void modifyWorldGenerator(WorldCreationSettings world, DataContainer settings, WorldGenerator worldGenerator) {
+        BiomeGenerationSettings desertSettings;
+        if(worldGenerator.isBiomeOverriden(BiomeTypes.DESERT)) {
+            desertSettings = worldGenerator.getBiomeOverride(BiomeTypes.DESERT);
+        } else {
+            desertSettings = BiomeTypes.DESERT.getGenerationSettings().clone();
+        }
+        for(Populator populator: desertSettings.getPopulators()) {
```
Comment (phroa) [inline source/plugin/wgen/customwgen.rst]: could probably Stream this.

File: source/plugin/wgen/customwgen.rst [hunk 6]
```diff
@@ -0,0 +1,240 @@
...
+Start by creating a new ``BiomeGenerationSettings`` instance which is a perfect copy of the settings specified
+by the desert biome. Then apply your changes to this instance, and finally apply this new instance to the world
+generator as an override for the desert biome.
+
+.. code-block:: java
+
+    @Override
+    public void modifyWorldGenerator(WorldCreationSettings world, DataContainer settings, WorldGenerator worldGenerator) {
+        BiomeGenerationSettings desertSettings;
+        if(worldGenerator.isBiomeOverriden(BiomeTypes.DESERT)) {
```
Comment (phroa) [inline source/plugin/wgen/customwgen.rst]: space after `if`

File: source/plugin/wgen/customwgen.rst [hunk 7]
```diff
@@ -0,0 +1,240 @@
...
+            desertSettings = worldGenerator.getBiomeOverride(BiomeTypes.DESERT);
+        } else {
+            desertSettings = BiomeTypes.DESERT.getGenerationSettings().clone();
+        }
+        for(Populator populator: desertSettings.getPopulators()) {
+            if(populator instanceof Cactus) {
```
Comment (phroa) [inline source/plugin/wgen/customwgen.rst]: space after `if`

File: source/plugin/wgen/customwgen.rst [hunk 8]
```diff
@@ -0,0 +1,240 @@
...
+
+Changing the base ``GenerationPopulator`` of a world generator allows you to change the base terrain shape
+generation of the world. A generator populator will roughly follow the procedure of using the seed and biome
+information to seed and modify a series of noise maps, from which the terrain is formed. The terrain created
+in a modified base generator populator should only consist of stone blocks, to allow the biomes to properly replace
+blocks for biome-specific ground cover.
+
+.. code-block:: java
+
+	public class SinusoidalGenerator implements GenerationPopulator {
```
Comment (phroa) [inline source/plugin/wgen/customwgen.rst]: more doubled indentation

File: source/plugin/wgen/customwgen.rst [hunk 9]
```diff
@@ -0,0 +1,240 @@
...
+			}
+		}
+
+		private int getHeight(int x, int z, BiomeType biome) {
+			double sx = Math.sin(x/64d)+1;
+			double sz = Math.sin(z/64d)+1;
+			double value = (sx + sz) / 4d;
+			BiomeGenerationSettings settings = biome.getGenerationSettings();
+			double heightRange = settings.getMaxHeight() - settings.getMinHeight();
+			double height = heightRange * value | settings.getMinHeight;
```
Comment (phroa) [inline source/plugin/wgen/customwgen.rst]: no parens after `getMinHeight`

File: source/plugin/wgen/customwgen.rst [hunk 10]
```diff
@@ -0,0 +1,240 @@
...
+
+    @Override
+    public void modifyWorldGenerator(WorldCreationSettings world, DataContainer settings, WorldGenerator worldGenerator) {
+        BiomeGenerationSettings desertSettings = BiomeTypes.DESERT.getGenerationSettings();
+        for(Populator populator: desertSettings.getPopulators()) {
+            if(populator instanceof Cactus) {
+                ((Cactus) populator).setHeight(5);
+            }
+        }
+    }
```
Comment (boformer) [inline source/plugin/wgen/customwgen.rst]: Imports for the code examples would be nice.

Comment (saladoc) [inline source/plugin/wgen/customwgen.rst]: done

**Comment by Inscrutable** [comment]
One more thing needs finishing, the **Creating Custom GenerationPopulators** section of the customwgen page is still TBA. Also, have the code snippets been tested?

**Comment by Deamon5550** [comment]
I'll just stub that section out for now until the genpop stuff is in API.

**Comment by Inscrutable** [comment]
A note that the API isn't yet ready for them would suffice. Then all that's left is to be sure the code examples are valid, and squash this ocean of commits. Over to you @Deamon5550

**Pull request merged**